### PR TITLE
Exhaustive examples

### PR DESCRIPTION
**Pull request by craigp-atw**
Add script to update the exhaustive examples for each concept by searching through all the source texts.

This also uses a new format for storing the examples as a series of these json objects:
```
{
   'reference': { 'type': 'Bible', 'id_primary': 'Genesis', 'id_secondary': '1', 'id_tertiary': '1' },
   'context': { [key: string]: [value: string] },
}
```

The keys in the 'context' generally match what is currently shown within TBTA, with some new ones for extra info.

### CODE REVIEW

**Comment by craigp-atw** [comment]
I'll be curious how the size of the db compares between the old and new formats.

File: exhaustive_examples/semantic_encoding.js
Comment (longrunningprocess) [inline exhaustive_examples/semantic_encoding.js:1]: do you foresee us ever needing/wanting to get this info from the database so it's data driven?

File: exhaustive_examples/index.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
+import { find_word_context } from './example_context'
+import { transform_semantic_encoding } from './semantic_encoding'
+import { Database } from 'bun:sqlite'
+
+// usage: `bun exhaustive_examples/index.js Ontology.VERSION.YYYY-MM-DD.tabitha.sqlite Sources.YYYY-MM-DD.tabitha.sqlite`
+const ontology_db_name	 = Bun.argv[2]	// Ontology.VERSION.YYYY-MM-DD.tabitha.sqlite
```
Comment (longrunningprocess) [inline exhaustive_examples/index.js:6]:
```suggestion
const ontology_db_name	 = Bun.argv[2]
```

File: exhaustive_examples/index.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+const sources_db_name = Bun.argv[3]	// Sources.YYYY-MM-DD.tabitha.sqlite
```
Comment (longrunningprocess) [inline exhaustive_examples/index.js:7]:
```suggestion
const sources_db_name = Bun.argv[3]
```

File: exhaustive_examples/example_context.js [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+/**
+ * 
+ * @param {(entity: SourceEntity) => boolean} entity_filter 
+ * @param {Object} [param1={}] 
+ * @param {boolean} [param1.skip_phrases=false] 
+ * @param {boolean} [param1.skip_clauses=false] 
+ * @param {(entity: SourceEntity) => boolean} [param1.break_condition=()=>false] 
+ * @returns {(start_index: number, source_entities: SourceEntity[]) => number}
+ */
+function find_entity_before(entity_filter, { skip_phrases=false, skip_clauses=false, break_condition=()=>false }={}) {
```
Comment (longrunningprocess) [inline exhaustive_examples/example_context.js:272]: this is pretty sophisticated bro 🧠

File: exhaustive_examples/example_context.js
Comment (longrunningprocess) [inline exhaustive_examples/example_context.js:1]: the functionality in here is intense but the code is clean, nice work!

File: exhaustive_examples/example_context.js [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+ */
+export function find_word_context(entity_index, source_entities) {
+	const entity = source_entities[entity_index]
+	return context_argument_finder[entity.label]?.(entity_index, source_entities) ?? {}
+}
+
+/**
+ * @type {Record<string, (entity_index: number, source_entities: SourceEntity[]) => ContextArguments>}
+ */
+const context_argument_finder = {
```
Comment (longrunningprocess) [inline exhaustive_examples/example_context.js:17]: this is really slick!

File: exhaustive_examples/index.js [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+		}
+
+		// For each word encountered, add the current verse reference to that word's examples
+		transform_semantic_encoding(semantic_encoding)
+			.reduce(group_examples_by_concept, new Map())
+			.forEach((context_arguments, concept_key) => {
+				const examples = context_arguments.map(context => JSON.stringify({ reference, context })).join('\n')
+				const { stem, sense, part_of_speech } = JSON.parse(concept_key)
+				
+				// TODO only update the concepts every chapter/book instead of every verse?
```
Comment (longrunningprocess) [inline exhaustive_examples/index.js:49]: does this still belong?

File: ontology/migrate_concepts_table.js [hunk 6]
```diff
@@ -26,11 +26,11 @@ function transform_tbta_data(tbta_db) {
 						Roots							AS stem,
 						''								AS sense,
 						'${part_of_speech}'		AS part_of_speech,
-						Occurrences					AS occurrences,
+						0					AS occurrences,
```
Comment (longrunningprocess) [inline ontology/migrate_concepts_table.js:29]:
```suggestion
						0						AS occurrences,
```

Comment (longrunningprocess) [inline ontology/migrate_concepts_table.js:29]:
```suggestion
						0								AS occurrences,
```

**Review by longrunningprocess — CHANGES_REQUESTED** [review]
code looks good, I'm going to run it locally and I'll let you know about the db size

File: exhaustive_examples/index.js
Comment (longrunningprocess) [inline exhaustive_examples/index.js:1]: how would you feel about renaming this file to `load.js`?

**Comment by longrunningprocess** [comment]
maybe we can chat about a failed local run:

```sh
bun exhaustive_examples/index.js Ontology.9488.2024-04-01.tabitha.sqlite Sources.2024-08-08.tabitha.sqlite
Fetching all source encoding...
Fetched 31482 verses

Collecting occurrences within 1 Chronicles:
..............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
Collecting occurrences within 1 Corinthians:
.....................................................................................................................................................................................................................................................................................................................................................................................................................................................
Collecting occurrences within 1 John:
.........................................................................................................
Collecting occurrences within 1 Kings:
.....................................................................................................................................................................................................................................................231 | 		entity => entity.label === word_type,
232 | 		{ skip_phrases: true, skip_clauses: true, break_condition: is_closing_phrase }
233 | 	)(phrase_index, source_entities)
234 |
235 | 	if (head_index === -1) {
236 | 		throw new Error(`Invalid semantic encoding - missing head ${word_type} in ${phrase_type}`)
              ^
error: Invalid semantic encoding - missing head Noun in NP
      at get_head_word (/Users/billyclark/projects/presciencelabs/tabitha-databases/exhaustive_examples/example_context.js:236:9)
      at value (/Users/billyclark/projects/presciencelabs/tabitha-databases/exhaustive_examples/example_context.js:65:35)
      at /Users/billyclark/projects/presciencelabs/tabitha-databases/exhaustive_examples/example_context.js:341:45
      at find_verb_context (/Users/billyclark/projects/presciencelabs/tabitha-databases/exhaustive_examples/example_context.js:84:6)
      at find_word_context (/Users/billyclark/projects/presciencelabs/tabitha-databases/exhaustive_examples/example_context.js:11:40)
      at group_examples_by_concept (/Users/billyclark/projects/presciencelabs/tabitha-databases/exhaustive_examples/index.js:87:28)
      at reduce (:1:11)
      at /Users/billyclark/projects/presciencelabs/tabitha-databases/exhaustive_examples/index.js:44:5

Bun v1.1.14 (macOS arm64)
```

File: exhaustive_examples/semantic_encoding.js
Comment (craigp-atw) [reply exhaustive_examples/semantic_encoding.js:1]: Yes, that would be best. See https://github.com/presciencelabs/tabitha-sources/issues/9

File: exhaustive_examples/index.js (hunk 5 above)
Comment (craigp-atw) [reply exhaustive_examples/index.js:49]: Removed

File: ontology/migrate_concepts_table.js (hunk 6 above)
Comment (craigp-atw) [reply ontology/migrate_concepts_table.js:29]: The spacing is all over the place for me, so I didn't bother touching it.
image

File: exhaustive_examples/index.js
Comment (craigp-atw) [reply exhaustive_examples/index.js:1]: sure

**Comment by craigp-atw** [comment]
We've changed the design for storing each example. Instead of a `examples` column in the concepts table that holds a large amount of text, a separate `Exhaustive_Examples` table contains a row for each example occurrence. The columns are
```
concept_stem,
concept_sense,
concept_part_of_speech,
ref_type,
ref_id_primary,
ref_id_secondary,
ref_id_tertiary,
context_json,
```

**Comment by craigp-atw** [comment]
> I'll be curious how the size of the db compares between the old and new formats.

The Ontology db with the old example format was ~6MB. The Ontology db with the new format is ~47MB. Keep in mind the examples are far more comprehensive now - 'person-A' stored ~1000 examples in the old format, whereas there are 30000+ in the new. And now the examples are much more searchable.

**Review by longrunningprocess — DISMISSED** [review]

**Pull request merged**